gax/src/main/java/com/google/api/gax/rpc/internal/ApiCallContextUtil.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc.internal;
+
+import com.google.api.core.InternalApi;
+import com.google.api.gax.rpc.ApiCallContext.Key;
+import com.google.common.collect.ImmutableMap;
+
+@InternalApi
+public final class ApiCallContextUtil {
```

Comment:
I think it would be nicer if encapsulated the data as well. Something like:

```java
class ApiCallContextOptions {
  <T> ApiOptions withOption(ApiCallContext.Key<T>, T) {
   ...
  }
  <T> T get(ApiCallContext.Key<T> key) {
  }
 ApiCallContextOptions merge(ApiCallContextOptions opts) {
 }
}
```

Then you can use it as an implementation detail in ApiCallContext impls. And be able to easily address my next question:

How come you didn't use grpc approach of linear search and used a hash map instead?

Reply:
Done. 

> How come you didn't use grpc approach of linear search and used a hash map instead?
`ApiCallContext#merge` will need to merge 2 contexts, so I think it's more efficient with hash map.